### PR TITLE
Bugfix CMR-10079: json access log

### PR DESCRIPTION
**Pull request by jceaser**
# Overview

The JSON Access log does not calculate duration correctly.

### What is the Solution?
The solution was to create a new Ring handler to capture the start time of a request. This handler is called as soon as possible in the Ring chain. Then the Log handler is to be called at the end of the chain and it uses the value to calculate duration.

While in the code, cleaned up the values populated:
* Removed values which are already in the headers
* Added all headers in bulk
* Added now-ms (current time as milliseconds)
* renamed uri to url
* added actual uri from system
* changed remote address so that IP6 and IP4 localhost addresses report as "localhost"

### What areas of the application does this impact?
Common
metadata-db, ingest, search, indexer, bootstrap, virtual-product, access-control

Not changed: mock-echo (port 2999)

A test was changed so as to not issue warnings locally about using command function names (parse-double).

# Checklist

- [x] I have updated/added unit and int tests that prove my fix is effective or that my feature works
- [x] New and existing unit and int tests pass locally and remotely
- [x] clj-kondo has been run locally and all errors corrected
- [x] I have removed unnecessary/dead code and imports in files I have changed
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [-] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [-] I have cleaned up integration tests by doing one or more of the following:
  - migrated any are2 tests to are3 in files I have changed
  - de-duped, consolidated, removed dead int tests
  - transformed applicable int tests into unit tests
  - refactored to reduce number of system state resets by updating fixtures (use-fixtures :each (ingest/reset-fixture {})) to be :once instead of :each

### CODE REVIEW

File: common-app-lib/src/cmr/common_app/api/request_logger.clj [hunk 1]
```diff
@@ -61,6 +65,26 @@
        ;; Parameters in forms could be really large, don't log everything
        (as-> item (apply dissoc item (drop limit (keys item)))))))
 
+(defn- normalize-local
+  "IP 6 addresses are ugly and wildly different then IP 4, normalize these so that all localhost
+   searches can be found in logs regardless of technology used to deliver the request."
+  [raw-address]
+  (if (contains? (set ["[0:0:0:0:0:0:0:1]" "127.0.0.1"]) raw-address)
+    "localhost"
+    raw-address))
```
Comment (jceaser) [inline common-app-lib/src/cmr/common_app/api/request_logger.clj:74]: if a system calls itself through transmit, or if running locally this comes into play. I also just felt that IP6 addresser were ugly.

File: common-app-lib/src/cmr/common_app/api/request_logger.clj [hunk 2]
```diff
@@ -61,6 +65,26 @@
...
+
+(defn- add-hash-to-data
```
Comment (jceaser) [inline common-app-lib/src/cmr/common_app/api/request_logger.clj:76]: only change here was to move it in the file and make it private

File: common-app-lib/src/cmr/common_app/api/request_logger.clj [hunk 3]
```diff
@@ -116,55 +136,60 @@
...
    (log-ring-request handler :ignore-id))
   ([handler id]
    (fn [request]
      (if-not (config/enable-enhanced-http-logging)
        (handler request)
-       (let [start (tk/now-ms)
+       (let [response (handler request) ;; Do all the response handlers first
+             log-start (tk/now-ms) ;; After processing the other handlers, start tracking log time
+             _ (when-not (:ring-start-time request) ;; Did someone forget to setup a routes.clj
+                 (error "There is no ring start time for this service" (request->uri request)))
```
Comment (jceaser) [inline common-app-lib/src/cmr/common_app/api/request_logger.clj:158]: when configured correctly, this error should never happen, but if someone changes the ring order, this message will be issued if the order results in a failed setup.

File: common-app-lib/src/cmr/common_app/api/request_logger.clj [hunk 4]
```diff
@@ -116,55 +136,60 @@
...
                       ;; assume that (add-body-hashes) has been run and reuse data
-                      (assoc "client-id" (get-in request [:headers "client-id"] "unknown")
+                      (into (:headers request)) ;; Bring in all the existing headers
+                      (assoc "cmr-hit" (get-in response [:headers "CMR-Hit"] "n/a")
+                             "cmr-took" (get-in response [:headers "CMR-Took"] "n/a")
                              "form-params" (dump-param form-params :form-params)
                              "method" (string/upper-case
                                        (name (get-in request [:request-method] "unknown")))
                              "now" now
-                             "protocol" (:protocol request)
```
Comment (jceaser) [inline common-app-lib/src/cmr/common_app/api/request_logger.clj:155]: you get this for free in headers

File: common-app-lib/src/cmr/common_app/api/request_logger.clj [hunk 5]
```diff
@@ -116,55 +136,60 @@
...
+                             "now-ms" log-start
```
Comment (jceaser) [inline common-app-lib/src/cmr/common_app/api/request_logger.clj:177]: now is a string, human readable, now-ms is a number you can do math with in splunk

File: system-int-test/src/cmr/system_int_test/data2/atom_json.clj [hunk 6]
```diff
@@ -52,17 +52,17 @@
   (fn [concept-type json-entry]
     concept-type))
 
-(defn parse-long
+(defn parse-long-local
```
Comment (jceaser) [inline system-int-test/src/cmr/system_int_test/data2/atom_json.clj:55]: local repel would complain that we were using a built in function name

**Review by isja17 — APPROVED** [review]

File: bootstrap-app/src/cmr/bootstrap/api/routes.clj [hunk 7]
```diff
@@ -31,7 +31,8 @@
    [ring.middleware.json :as ring-json]
    [ring.middleware.keyword-params :as keyword-params]
    [ring.middleware.nested-params :as nested-params]
-   [ring.middleware.params :as params]))
+   [ring.middleware.params :as params]
+   [cmr.common-app.api.request-logger :as req-log]))
```
Comment (eereiter) [inline bootstrap-app/src/cmr/bootstrap/api/routes.clj:35]: This should be moved up in alphabetical order.

File: virtual-product-app/src/cmr/virtual_product/api/routes.clj [hunk 8]
```diff
@@ -1,25 +1,26 @@
...
+   [compojure.route :as route]
+   [ring.middleware.json :as ring-json]
+   [cmr.common-app.api.health :as common-health]
+   [cmr.common-app.api.request-logger :as req-log]
+   [cmr.common-app.api.routes :as common-routes]
+   [cmr.common.api.context :as context]
+   [cmr.common.api.errors :as errors]
+   [cmr.common.mime-types :as mt]
+   [cmr.virtual-product.services.health-service :as hs]
+   [compojure.handler :as handler]))
```
Comment (eereiter) [inline virtual-product-app/src/cmr/virtual_product/api/routes.clj:15]: If you are going to fix these they should be in alphabetical order.

**Review by eereiter — APPROVED** [review]

File: bootstrap-app/src/cmr/bootstrap/api/routes.clj (hunk 7 above)
Comment (jceaser) [reply bootstrap-app/src/cmr/bootstrap/api/routes.clj:35]: I also ran condo on the file and cleaned up other things.

**Pull request merged**